src/analyses/ai.cpp

```diff
@@ -327,118 +316,174 @@ bool ai_baset::visit_edge(
-      ns,
-      working_set);
+    if(callee_expression.id() == ID_symbol)
+    {
+      const irep_idt &callee_function_id =
+        to_symbol_expr(callee_expression).get_identifier();
+
+      goto_functionst::function_mapt::const_iterator it =
+        goto_functions.function_map.find(callee_function_id);
+
+      DATA_INVARIANT(
+        it != goto_functions.function_map.end(),
+        "Function " + id2string(callee_function_id) + "not in function map");
+
+      const goto_functionst::goto_functiont &callee_fun = it->second;
+
+      if(callee_fun.body_available())
+      {
+        return visit_edge_function_call(
+          calling_function_id,
+          l_call,
+          l_return,
+          callee_function_id,
+          working_set,
+          callee_fun.body,
+          goto_functions,
+          ns);
+      }
+      else
+      {
+        // Fall through to the default, body not available, case
+      }
+    }
+    else
+    {
+      // Function pointers are not currently supported and must be removed
+      DATA_INVARIANT(
+        callee_expression.id() == ID_symbol,
+        "Function pointers and indirect calls must be removed before "
+        "analysis.");
+    }
   }
 
-  DATA_INVARIANT(
-    !goto_function.body.instructions.empty(),
-    "By analysis time, all functions should have bodies");
+  // If the body is not available then we just do the edge from call to return
+  // in the caller.  Domains should over-approximate what the function might do.
+  return visit_edge(
+    calling_function_id,
+    l_call,
+    calling_function_id,
+    l_return,
+    ns,
+    working_set);
+}
 
-  // This is the edge from call site to function head.
+bool ai_baset::visit_end_function(
+  const irep_idt &function_id,
+  locationt l,
+  working_sett &working_set,
+  const goto_programt &goto_program,
+  const goto_functionst &goto_functions,
+  const namespacet &ns)
+{
+  PRECONDITION(l->is_end_function());
+
+  // Do nothing
+  return false;
+}
+
+void ai_baset::fixedpoint(
+  const goto_functionst &goto_functions,
+  const namespacet &ns)
+{
+  goto_functionst::function_mapt::const_iterator
+    f_it=goto_functions.function_map.find(goto_functions.entry_point());
+
+  if(f_it!=goto_functions.function_map.end())
+    fixedpoint(f_it->first, f_it->second.body, goto_functions, ns);
+}
 
+bool ai_recursive_interproceduralt::visit_edge_function_call(
+  const irep_idt &calling_function_id,
+  locationt l_call,
+  locationt l_return,
+  const irep_idt &callee_function_id,
+  working_sett &working_set,
+  const goto_programt &callee,
+  const goto_functionst &goto_functions,
+  const namespacet &ns)
+{
+  // This is the edge from call site to function head.
   {
-    locationt l_begin=goto_function.body.instructions.begin();
+    locationt l_begin = callee.instructions.begin();
 
-    working_sett working_set; // Redundant; fixpoint will add l_begin
+    working_sett ignore_working_set; // Redundant; fixpoint will add l_begin
```

Comment:
Up until the last commit of the PR the fixpoint function has a design oddity; it starts the fixpoint calculation with the first element of the function.  What this means is that the entry in the working set that is generated by computing the call edge will be ignored and the same thing recreated when the recursive call to fixpoint is made.  The naming of the variable is intended to make that a little clearer.

The last patch adds a 'starting history' parameter to fixpoint so that history objects can be threaded through calls in the expected way rather than being destroyed and recreated.  So both the code and the comment should be gone by the time the PR is merged.